react/features/app/getRouteToRender.js

```diff
@@ -49,60 +49,71 @@ export function _getRouteToRender(stateful: Function | Object): Promise<Route> {
-                    route.href = locationURL.href;
-                    resolve(route);
-                } else {
-                    // Maybe show deep-linking, otherwise go to Conference.
-                    getDeepLinkingPage(state).then(component => {
-                        route.component = component || Conference;
-                        resolve(route);
-                    });
-                }
-            }
+        return Promise.resolve(route);
+    }
 
-            return;
-        }
+    // Web custom routing when trying to join a meeting. There are essentially
+    // two possible handled routes here: welcome page and conference.
 
-        if (!isWelcomePageUserEnabled(state)) {
-            // Web: if the welcome page is disabled, go directly to a random
-            // room.
+    // We are intentionally not performing the check for mobile browsers because:
+    // - the WelcomePage is mobile ready;
+    // - if the URL points to a conference, getDeepLinkingPage will take
+    //   care of it.
+    const isUnsupportedBrowser
+        = !isMobileBrowser && !JitsiMeetJS.isWebRtcSupported();
 
-            let href = window.location.href;
+    if (joiningValidRoom) {
+        // Update the location if it doesn't match. This happens when a room is
+        // joined from the welcome page. The reason for doing this instead of
+        // using the history API is that we want to load the config.js which
+        // takes the room into account.
+        const { locationURL } = state['features/base/connection'];
 
-            href.endsWith('/') || (href += '/');
-            route.href = href + generateRoomWithoutSeparator();
-        } else if (isWelcomePageAppEnabled(state)) {
-            // Mobile: only go to the welcome page if enabled.
+        if (window.location.href !== locationURL.href) {
+            route.href = locationURL.href;
+
+            return Promise.resolve(route);
+        }
+
+        return getDeepLinkingPage(state)
+            .then(component => {
+                if (component) {
+                    route.component = component;
+                } else if (isUnsupportedBrowser) {
+                    route.component = UnsupportedDesktopBrowser;
+                } else {
+                    route.component = Conference;
+                }
 
+                return Promise.resolve(route);
+            });
+    }
+
+    // Web custom routing when trying to view the welcome page
+    if (isWelcomePageUserEnabled(state)) {
+        if (isUnsupportedBrowser) {
+            route.component = UnsupportedDesktopBrowser;
+        } else {
             route.component = WelcomePage;
         }
+    } else {
+        // Web: if the welcome page is disabled, go directly to a random room.
+
+        let href = window.location.href;
+
+        href.endsWith('/') || (href += '/');
+        route.href = href + generateRoomWithoutSeparator();
+    }
 
-        resolve(route);
-    });
+    return Promise.resolve(route);
```

Comment:
this method almost feels like it could be ~3 helper methods, but maybe it doesn't split nicely

Reply:
If the new logic is sound, and addresses the same flows as existing logic, then I can see this being split into three helpers, yes. I'll give it a try on a commit on top of the existing changes.